proposals/003-metric-improvenents.md

```diff
@@ -0,0 +1,262 @@
+much latency is being added for their particular use-cases and have the ability to monitor this latency over time.
+
+Kroxylicious Filters themselves have the potential to alter behaviour of the Kafka system.  Users will need insights into
+performance and other issues that might result from filter behaviour. Specifically:
+
+* filters have the ability to change the size of the RPCs as they traverse the proxy (record encryption)
+* filters have the ability to change the flow of messages within the Kafka system. Specifically, they can short-circuit
+  requests by producing their own responses (the broker never sees the request) and introduce their own requests.
+* filters can perform asynchronous work.  When asynchronous work occurs, the proxy stops reading the downstream or upstream.
+
+## Proposal
+
+## Summary
+
+The following metrics will be added.  They will each be described in more detail later in this section.
+
+| Metric                                           | Type         | Labels                                                                                | Description                                                                             |
+|--------------------------------------------------|--------------|---------------------------------------------------------------------------------------|-----------------------------------------------------------------------------------------|
+| kroxylicious_downstream_kafka_message_total      | Counter      | virtual_cluster, node_id†, flowing, api_key, api_version, opaque                      | Incremented by one every time a request comes from/response goes to the downstream.     |
+| kroxylicious_upstream_kafka_message_total        | Counter      | virtual_cluster, node_id†, flowing, api_key, api_version, opaque, originating_filter† | Incremented by one every time a request goes to/response comes from the upstream.       |
+| kroxylicious_downstream_kafka_message_size_bytes | Distribution | virtual_cluster, node_id†, flowing, api_key, api_version  opaque                      | Records the message size of every message coming from/response going to the downstream. |
+| kroxylicious_upstream_kafka_message_size_bytes   | Distribution | virtual_cluster, node_id†, flowing, api_key, api_version, opaque, originating_filter† | Records the message size of every message coming from/response going to the upstream.   |
+| kroxylicious_message_process_time                | Distribution | virtual_cluster, node_id†, flowing, api_key, api_version, opaque, originating_filter† | Records the time taken for the message to traverse the proxy                            |
+| kroxylicious_downstream_backpressure_state       | Gauge        | virtual_cluster, node_id†                                                             | Records the backpressure state from the downstream                                      |
```

Comment:
```suggestion
| kroxylicious_downstream_backpressure_state       | Gauge        | virtual_cluster, node_id†                                                             | Records the backpressure state being applied to the downstream                                      |
```

But more importantly, backpressure is a per-connection thing, but this is a Gauge, and there is no label for the client's identity, so I'm confused.

Reply:
Having a client_id, principal or channel as a label would threaten a cardinality explosion.

The use case for this is to get a broad understanding of how often the proxy is toggling autoread.  You are right, it won't be sufficient to identify root cause but it should be sufficient to tell an investigation if they are on the right track.  I'd expect a user with an auto read problem to turn to tracing or filter specific metrics to delve deeper.